google/cloud/storage/internal/minimal_iam_credentials_rest.cc

```diff
@@ -0,0 +1,151 @@
+#include "google/cloud/storage/internal/curl_handle_factory.h"
+#include "google/cloud/storage/internal/curl_request_builder.h"
+#include "google/cloud/storage/oauth2/credentials.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/absl_str_join_quiet.h"
+#include "google/cloud/internal/api_client_header.h"
+#include "google/cloud/internal/format_time_point.h"
+#include "google/cloud/internal/parse_rfc3339.h"
+#include "google/cloud/log.h"
+#include <nlohmann/json.hpp>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+
+namespace {
+
+class MinimalIamCredentialsRestImpl : public MinimalIamCredentialsRest {
+ public:
+  explicit MinimalIamCredentialsRestImpl(
+      std::shared_ptr<oauth2::Credentials> credentials, Options options)
+      : endpoint_(NormalizeEndpoint(options.get<RestEndpointOption>())),
+        credentials_(std::move(credentials)),
+        handle_factory_(std::make_shared<DefaultCurlHandleFactory>(options)),
+        x_goog_api_client_header_("x-goog-api-client: " + x_goog_api_client()),
+        options_(std::move(options)) {}
+
+  StatusOr<google::cloud::internal::AccessToken> GenerateAccessToken(
+      GenerateAccessTokenRequest const& request) override {
+    auto auth_header = credentials_->AuthorizationHeader();
+    if (!auth_header) return std::move(auth_header).status();
+
+    CurlRequestBuilder builder(MakeRequestUrl(request), handle_factory_);
+    builder.SetMethod("POST")
+        .ApplyClientOptions(options_)
+        .AddHeader(auth_header.value())
+        .AddHeader(x_goog_api_client_header_);
+    builder.AddHeader("Content-Type: application/json");
+    nlohmann::json payload{
+        {"delegates", request.delegates},
+        {"scope", request.scopes},
+        {"lifetime", std::to_string(request.lifetime.count()) + "s"},
+    };
+    auto response = builder.BuildRequest().MakeRequest(payload.dump());
+    if (!response) return std::move(response).status();
+    if (response->status_code >= HttpStatusCode::kMinNotSuccess) {
+      return AsStatus(*response);
+    }
+    auto parsed = nlohmann::json::parse(response->payload, nullptr);
+    if (parsed.is_null() || parsed.count("accessToken") == 0 ||
+        parsed.count("expireTime") == 0) {
+      return Status(StatusCode::kUnknown,
+                    "invalid response from service <" + parsed.dump() + ">");
+    }
+    return google::cloud::internal::AccessToken{
+        parsed["accessToken"].get<std::string>(),
+        google::cloud::internal::ParseRfc3339(
+            parsed["expireTime"].get<std::string>())};
+  }
+
+ private:
+  std::string MakeRequestUrl(GenerateAccessTokenRequest const& request) {
+    return absl::StrCat(endpoint_, "projects/-/serviceAccounts/",
+                        request.service_account, ":generateAccessToken");
+  }
+  static std::string NormalizeEndpoint(std::string endpoint) {
```

Comment:
This function could probably be moved to the .cc file in an unnamed namespace. I don't see any need for it to be in a header.

Reply:
I tricked you: it already is in the `.cc` file... the whole class is in the anonymous namespace in the `.cc` file.. :wink:

Reply:
D'oh! You got me :-)